### ISSUE TITLE
Ash.Changeset generates nil value within `attributes` when input is an empty string

### ISSUE THREAD

**Issue by moxley** [ash-project/ash#781]
**Describe the bug**

When calling `Ash.Changeset.for_create(resource, action, %{text: ""})`, the resulting changeset has an `:attributes` value of `%{text: nil}`. The expected value is `%{text: ""}`.

When this changeset is used to create a record with AshPostgres, the returned struct has the `nil` value instead of the expected empty string. The database record has the expected empty string.

This bug a likely the cause of this [ash_postgres bug](https://github.com/ash-project/ash_postgres/issues/186).

**To Reproduce**

```elixir
defmodule GF.Groups.MembershipPlan do
  use Ash.Resource,
    data_layer: AshPostgres.DataLayer,
    extensions: [AshGraphql.Resource],
    authorizers: [Ash.Policy.Authorizer]

  postgres do
    table "membership_plans"
    repo Repo
  end

  attributes do
    uuid_primary_key :id
    # It doesn't matter whether there is a default specified or not
    attribute :text, :string, default: ""
  end
```

Test 1:
```elixir
    test "attributes" do
      changeset =
        GF.Groups.MembershipPlan
        |> Ash.Changeset.for_create(:create, %{text: ""},
          authorize?: false
        )

      # Fails: actual value is %{text: nil}
      assert %{text: ""} = changeset.attributes
```

Test 2:
```elixir
    test "create" do
      plan =
        GF.Groups.MembershipPlan
        |> Ash.Changeset.for_create(:create, %{text: ""},
          authorize?: false
        )
        |> GF.Ash.Api.create!()

      # Fails: actual value is nil
      assert plan.text == ""
```

**Expected behavior**

Expected value to be an empty string.

**Runtime**
 - Elixir version: 1.14.3-otp-25
 - Erlang version: 25.2.2
 - OS: MacOS 14.1.1
 - Ash version: 2.17.4
 - any related extension versions: ash_postgres 1.3.41

**Additional context**

**Comment by moxley** [comment]
Turns out this is not a bug. Here's what I needed to do to fix this:

```elixir
defmodule GF.Groups.MembershipPlan do
  ...
  attributes do
    ...
    attribute :text, :string, default: "", constraints: [allow_empty?: true]
  end
```

`constraints: [allow_empty?: true]` is required to allow an empty string for this field.

**Comment by moissela** [comment]
Hi,
today I encountered the issue discussed here (using Ash v2.21.6) and I noticed that adding the constraint alone is not enough: it is necessary to also specify a default for the attribute like @moxley said.

Is this the desired behavior @zachdaniel?

I think it might be useful to allow saving an empty string even without specifying a default for the attribute.
This way passing to a changeset an empty string value will have a different results from passing a nil value or not passing a value at all.
